File: web/scripts/build-templates.ts
// 正解データ (../data/results_gt) から数字テンプレートを作り public/data/digit-templates.json に保存する。
// usage: npm run templates
import { writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { DigitReader, type DigitField, FIELDS } from "../src/core/digits";
import { DATA, digitsText, loadGt, loadImage } from "./node-env";

const reader = new DigitReader();
let skipped = 0;
for (const r of loadGt()) {
  const img = loadImage(`${DATA}/results_raw/${r.file}`);
  for (const key of Object.keys(FIELDS) as DigitField[]) {
    if (r[key] && Number(r[key]) > 0 && !reader.add(img, key, digitsText(key, r[key]), r.file)) skipped++;
  }
}
const out = resolve(import.meta.dirname, "../public/data/digit-templates.json");
writeFileSync(out, JSON.stringify(reader.toJSON()));
console.log(`score ${reader.templates.score.length} / count ${reader.templates.count.length} (切り出し数不一致 ${skipped}) -> ${out}`);

File: scripts/build_templates.py
"""正解データ (data/results_gt) から数字テンプレートを作り data/digit_templates.npz に保存する。

usage: python scripts/build_templates.py
"""
import os
import sys

import cv2

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, ROOT)
from core.digits import FIELDS, DigitReader, as_text, load_gt

if __name__ == "__main__":
    reader = DigitReader()
    for r in load_gt():
        img = cv2.imread(os.path.join(ROOT, "data/results_raw", r["file"]))
        for key in FIELDS:
            if r[key] and int(r[key]) > 0:
                reader.add(img, key, as_text(key, r[key]), r["file"])
    out = os.path.join(ROOT, "data/digit_templates.npz")
    reader.save(out)
    print({font: len(t) for font, t in reader.templates.items()}, "->", out)
